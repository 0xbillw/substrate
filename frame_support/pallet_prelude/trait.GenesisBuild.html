<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="A trait to define the build function of a genesis config, T and I are placeholder for pallet trait and pallet instance."><meta name="keywords" content="rust, rustlang, rust-lang, GenesisBuild"><title>GenesisBuild in frame_support::pallet_prelude - Rust</title><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../../ayu.css" disabled ><script id="default-settings" ></script><script src="../../storage.js"></script><script src="../../crates.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../../favicon.svg"><link rel="alternate icon" type="image/png" href="../../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><script defer type="text/javascript" src="https://substrate.dev/js/config.js"></script>
<script defer type="text/javascript" src="https://substrate.dev/js/klaro.min.js"></script>
<script async type="opt-in" data-type="application/javascript"
		data-src="https://www.googletagmanager.com/gtag/js?id=UA-145158313-2" data-name="googleAnalytics"></script>
<script type="opt-in" data-type="application/javascript" data-name="googleAnalytics">
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());
      gtag('config', 'UA-145158313-2');
</script>

<style type="text/css">#crate-search{background-image:url("../../down-arrow.svg");}</style></head><body class="rustdoc trait"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../../frame_support/index.html'><div class='logo-container rust-logo'><img src='../../rust-logo.png' alt='logo'></div></a><p class="location">Trait GenesisBuild</p><div class="sidebar-elems"><div class="block items"><a class="sidebar-title" href="#required-methods">Required Methods</a><div class="sidebar-links"><a href="#tymethod.build">build</a></div><a class="sidebar-title" href="#provided-methods">Provided Methods</a><div class="sidebar-links"><a href="#method.assimilate_storage">assimilate_storage</a><a href="#method.build_storage">build_storage</a></div><a class="sidebar-title" href="#implementors">Implementors</a></div><p class="location">Other items in<br><a href="../index.html">frame_support</a>::<wbr><a href="index.html">pallet_prelude</a></p><div id="sidebar-vars" data-name="GenesisBuild" data-ty="trait" data-relpath=""></div><script defer src="sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img width="18" height="18" alt="Pick another theme!" src="../../brush.svg"></button><div id="theme-choices" role="menu"></div></div><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input"name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../../settings.html" title="settings"><img width="18" height="18" alt="Change settings" src="../../wheel.svg"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Trait <a href="../index.html">frame_support</a>::<wbr><a href="index.html">pallet_prelude</a>::<wbr><a class="trait" href="#">GenesisBuild</a><button id="copy-path" onclick="copy_path(this)" title="copy path"><img src="../../clipboard.svg" width="19" height="18" alt="Copy item import" title="Copy item import to clipboard"></button></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../../src/frame_support/traits/hooks.rs.html#304-323" title="goto source code">[src]</a></span></h1><div class="docblock type-decl"><pre class="rust trait">pub trait GenesisBuild&lt;T, I&nbsp;=&nbsp;<a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.unit.html">()</a>&gt;: <a class="trait" href="https://doc.rust-lang.org/nightly/core/default/trait.Default.html" title="trait core::default::Default">Default</a> + <a class="trait" href="trait.MaybeSerializeDeserialize.html" title="trait frame_support::pallet_prelude::MaybeSerializeDeserialize">MaybeSerializeDeserialize</a> {
    fn <a href="#tymethod.build" class="fnname">build</a>(&amp;self);

    fn <a href="#method.build_storage" class="fnname">build_storage</a>(&amp;self) -&gt; <a class="enum" href="../dispatch/result/enum.Result.html" title="enum frame_support::dispatch::result::Result">Result</a>&lt;<a class="struct" href="../../sp_storage/struct.Storage.html" title="struct sp_storage::Storage">Storage</a>, <a class="struct" href="https://doc.rust-lang.org/nightly/alloc/string/struct.String.html" title="struct alloc::string::String">String</a>&gt; { ... }
<div class="item-spacer"></div>    fn <a href="#method.assimilate_storage" class="fnname">assimilate_storage</a>(&amp;self, storage: &amp;mut <a class="struct" href="../../sp_storage/struct.Storage.html" title="struct sp_storage::Storage">Storage</a>) -&gt; <a class="enum" href="../dispatch/result/enum.Result.html" title="enum frame_support::dispatch::result::Result">Result</a>&lt;<a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.unit.html">()</a>, <a class="struct" href="https://doc.rust-lang.org/nightly/alloc/string/struct.String.html" title="struct alloc::string::String">String</a>&gt; { ... }
}</pre></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>A trait to define the build function of a genesis config, T and I are placeholder for pallet
trait and pallet instance.</p>
</div></details><h2 id="required-methods" class="small-section-header">Required methods<a href="#required-methods" class="anchor"></a></h2><div class="methods"><details class="rustdoc-toggle" open><summary><div id="tymethod.build" class="method has-srclink"><div class="rightside"><a class="srclink" href="../../src/frame_support/traits/hooks.rs.html#307" title="goto source code">[src]</a></div><code>fn <a href="#tymethod.build" class="fnname">build</a>(&amp;self)</code></div></summary><div class="docblock"><p>The build function is called within an externalities allowing storage APIs.
Thus one can write to storage using regular pallet storages.</p>
</div></details></div><h2 id="provided-methods" class="small-section-header">Provided methods<a href="#provided-methods" class="anchor"></a></h2><div class="methods"><details class="rustdoc-toggle" open><summary><div id="method.build_storage" class="method has-srclink"><div class="rightside"><a class="srclink" href="../../src/frame_support/traits/hooks.rs.html#310-314" title="goto source code">[src]</a></div><code>fn <a href="#method.build_storage" class="fnname">build_storage</a>(&amp;self) -&gt; <a class="enum" href="../dispatch/result/enum.Result.html" title="enum frame_support::dispatch::result::Result">Result</a>&lt;<a class="struct" href="../../sp_storage/struct.Storage.html" title="struct sp_storage::Storage">Storage</a>, <a class="struct" href="https://doc.rust-lang.org/nightly/alloc/string/struct.String.html" title="struct alloc::string::String">String</a>&gt;</code></div></summary><div class="docblock"><p>Build the storage using <code>build</code> inside default storage.</p>
</div></details><details class="rustdoc-toggle" open><summary><div id="method.assimilate_storage" class="method has-srclink"><div class="rightside"><a class="srclink" href="../../src/frame_support/traits/hooks.rs.html#317-322" title="goto source code">[src]</a></div><code>fn <a href="#method.assimilate_storage" class="fnname">assimilate_storage</a>(&amp;self, storage: &amp;mut <a class="struct" href="../../sp_storage/struct.Storage.html" title="struct sp_storage::Storage">Storage</a>) -&gt; <a class="enum" href="../dispatch/result/enum.Result.html" title="enum frame_support::dispatch::result::Result">Result</a>&lt;<a class="primitive" href="https://doc.rust-lang.org/nightly/std/primitive.unit.html">()</a>, <a class="struct" href="https://doc.rust-lang.org/nightly/alloc/string/struct.String.html" title="struct alloc::string::String">String</a>&gt;</code></div></summary><div class="docblock"><p>Assimilate the storage for this module into pre-existing overlays.</p>
</div></details></div><h2 id="implementors" class="small-section-header">Implementors<a href="#implementors" class="anchor"></a></h2><div class="item-list" id="implementors-list"></div><script type="text/javascript" src="../../implementors/frame_support/pallet_prelude/trait.GenesisBuild.js" async></script></section><section id="search" class="content hidden"></section><div id="rustdoc-vars" data-root-path="../../" data-current-crate="frame_support" data-search-index-js="../../search-index.js" data-search-js="../../search.js"></div>
    <script src="../../main.js"></script>
</body></html>