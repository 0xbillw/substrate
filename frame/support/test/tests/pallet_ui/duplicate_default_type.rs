#[frame_support::pallet]
mod pallet {
	use frame_support::pallet_prelude::*;

	#[pallet::config]
	pub trait Config: frame_system::Config {
        #[pallet::default_type(u64)]
        #[pallet::default_type(u128)]
        type Balance: Parameter;
    }

	#[pallet::pallet]
	#[pallet::generate_store(trait Store)]
	pub struct Pallet<T>(core::marker::PhantomData<T>);
}

fn main() {
}
