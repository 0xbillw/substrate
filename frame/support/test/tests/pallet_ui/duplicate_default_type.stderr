error: Invalid attribute in pallet::config, only one pallet::default_type is expected
 --> $DIR/duplicate_default_type.rs:8:19
  |
8 |         #[pallet::default_type(u128)]
  |                   ^^^^^^^^^^^^
